File: src/main/java/thkoeln/dungeon/eventconsumer/core/AbstractEvent.java
package thkoeln.dungeon.eventconsumer.core;

import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.messaging.MessageHeaders;
import thkoeln.dungeon.player.application.PlayerApplicationService;

import javax.persistence.Id;
import javax.persistence.MappedSuperclass;
import javax.persistence.Transient;
import java.util.UUID;

@MappedSuperclass
@Getter
@Setter
@NoArgsConstructor( access = AccessLevel.PROTECTED )
public abstract class AbstractEvent {
    @Id
    @Setter( AccessLevel.NONE )
    protected UUID id = UUID.randomUUID();
    protected UUID eventId;
    protected Long timestamp;
    protected UUID transactionId;

    @Transient
    protected Logger logger = LoggerFactory.getLogger(PlayerApplicationService.class);

    private static final String TRANSACTION_ID_KEY = "transactionId";

    public AbstractEvent( MessageHeaders messageHeaders ) {
        setEventId( messageHeaders.getId() );
        setTimestamp( messageHeaders.getTimestamp() ) ;
        try {
            setTransactionId(UUID.fromString(String.valueOf(messageHeaders.get(TRANSACTION_ID_KEY))));
        }
        catch ( IllegalArgumentException e ) {
            logger.warn( "Event " + eventId + " at time " + timestamp + " doesn't have a transactionId." );
        }
    }
}

File: src/main/java/thkoeln/dungeon/eventconsumer/core/AbstractEvent.kt
package thkoeln.dungeon.eventconsumer.core

import lombok.*
import org.slf4j.LoggerFactory
import org.springframework.messaging.MessageHeaders
import thkoeln.dungeon.player.application.PlayerApplicationService
import java.util.*
import javax.persistence.Id
import javax.persistence.MappedSuperclass
import javax.persistence.Transient

@MappedSuperclass
@Getter
@Setter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
abstract class AbstractEvent(messageHeaders: MessageHeaders) {
    @Id
    @Setter(AccessLevel.NONE)
    protected var id = UUID.randomUUID()

    protected var eventId: UUID? = null
    protected var timestamp: Long? = null
    protected var transactionId: UUID? = null

    @Transient
    protected var logger = LoggerFactory.getLogger(PlayerApplicationService::class.java)

    init {
        this.eventId = messageHeaders.id
        this.timestamp = messageHeaders.timestamp
        try {
            this.transactionId = UUID.fromString(messageHeaders[TRANSACTION_ID_KEY].toString())
        } catch (e: IllegalArgumentException) {
            logger.warn("Event $eventId at time $timestamp doesn't have a transactionId.")
        }
    }

    companion object {
        private const val TRANSACTION_ID_KEY = "transactionId"
    }
}